File: frontend/apps/ios/Pantopus/Features/Homes/VerifyLandlord/Postcard/PostcardVerificationViewModel.swift
//
//  PostcardVerificationViewModel.swift
//  Pantopus
//
//  A12.7 — sibling status screen for the verify-landlord flow. Tracks
//  the postcard's USPS state (mailed / in transit / delivered) and the
//  user's 6-char code input. The submit state machine is the same
//  contract shared with the wizard VM: idle → submitting → submitted /
//  error.
//

import Foundation
import Observation

/// Lifecycle state of the physical postcard.
public enum PostcardDeliveryStage: Sendable, Equatable {
    case mailed
    case inTransit
    case delivered
}

/// Full payload describing the postcard verification surface.
public struct PostcardVerificationContent: Sendable, Equatable {
    public let recipientName: String
    public let street: String
    public let cityZip: String
    public let trackingNumber: String
    public let mailedOn: String
    public let inTransitOn: String?
    public let deliveredOn: String?
    public let resendAvailableOn: String

    public init(
        recipientName: String,
        street: String,
        cityZip: String,
        trackingNumber: String,
        mailedOn: String,
        inTransitOn: String?,
        deliveredOn: String?,
        resendAvailableOn: String
    ) {
        self.recipientName = recipientName
        self.street = street
        self.cityZip = cityZip
        self.trackingNumber = trackingNumber
        self.mailedOn = mailedOn
        self.inTransitOn = inTransitOn
        self.deliveredOn = deliveredOn
        self.resendAvailableOn = resendAvailableOn
    }
}

/// Deterministic sample data — used by previews + snapshot tests, and
/// as the default seed when a homeId pattern doesn't match a stored
/// postcard.
public enum PostcardVerificationSampleData {
    public static let deliveredContent = PostcardVerificationContent(
        recipientName: "Mira Patel",
        street: "412 Elm St, Apt 3B",
        cityZip: "San Francisco, CA 94114",
        trackingNumber: "#9405 5036 …8421",
        mailedOn: "Oct 9",
        inTransitOn: "Oct 11",
        deliveredOn: "Oct 12",
        resendAvailableOn: "Oct 15"
    )

    public static let inTransitContent = PostcardVerificationContent(
        recipientName: "Mira Patel",
        street: "412 Elm St, Apt 3B",
        cityZip: "San Francisco, CA 94114",
        trackingNumber: "#9405 5036 …8421",
        mailedOn: "Oct 9",
        inTransitOn: "Oct 11",
        deliveredOn: nil,
        resendAvailableOn: "Oct 15"
    )

    public static func content(for stage: PostcardDeliveryStage) -> PostcardVerificationContent {
        stage == .delivered ? deliveredContent : inTransitContent
    }

    public static func stage(for homeId: String) -> PostcardDeliveryStage {
        if homeId.localizedCaseInsensitiveContains("delivered") {
            return .delivered
        }
        return .inTransit
    }
}

/// Outbound events the host nav stack acts on.
public enum PostcardVerificationOutboundEvent: Sendable, Equatable {
    case dismiss
    /// Verify pressed and the code matched — caller should pop the
    /// screen and route to the verified-home success surface. The
    /// homeId is forwarded so callers can refresh that home's
    /// verification status.
    case verified(homeId: String)
}

/// View model for A12.7. Holds the current delivery stage, the 6-char
/// code the user is typing, and a `submitState` machine identical in
/// shape to the wizard's `VerifyLandlordSubmitState`.
@Observable
@MainActor
final class PostcardVerificationViewModel {
    // MARK: - Published state

    private(set) var stage: PostcardDeliveryStage
    private(set) var content: PostcardVerificationContent
    var codeInput: String = ""
    private(set) var submitState: VerifyLandlordSubmitState = .idle
    var pendingEvent: PostcardVerificationOutboundEvent?

    // MARK: - Init

    private let homeId: String
    private let submitDelayNanos: UInt64
    private let api: APIClient
    /// Offline/preview/test seam. When non-nil, `verify()` checks the
    /// typed code against this value locally instead of calling the
    /// backend — used by previews and unit/snapshot tests. Production
    /// passes `nil`, so the code is validated by
    /// `POST /api/homes/:id/verify-postcard`.
    private let expectedCode: String?

    init(
        homeId: String,
        stage: PostcardDeliveryStage? = nil,
        content: PostcardVerificationContent? = nil,
        expectedCode: String? = nil,
        api: APIClient = .shared,
        submitDelayNanos: UInt64 = 800_000_000
    ) {
        let resolvedStage = stage ?? PostcardVerificationSampleData.stage(for: homeId)
        self.stage = resolvedStage
        self.content = content
            ?? PostcardVerificationSampleData.content(for: resolvedStage)
        self.homeId = homeId
        self.expectedCode = expectedCode
        self.api = api
        self.submitDelayNanos = submitDelayNanos
    }

    // MARK: - Derived state

    /// Whether the user can type into the 6-char field. Locked while
    /// in transit so a wrong-code submit can't happen before delivery.
    var isCodeInputUnlocked: Bool {
        stage == .delivered
    }

    var isSubmitting: Bool {
        if case .submitting = submitState { return true }
        return false
    }

    /// Whether the primary CTA fires. Mirrors the design's disabled
    /// state on the in-transit frame.
    var primaryCTAEnabled: Bool {
        stage == .delivered && codeInput.count == 6 && !isSubmitting
    }

    var primaryCTALabel: String {
        isSubmitting ? "Verifying…" : "Verify code"
    }

    // MARK: - Mutations

    func updateCode(_ raw: String) {
        codeInput = String(raw.uppercased().prefix(6))
    }

    /// Resend the postcard — reissues a new code via
    /// `POST /api/homes/:id/request-postcard`. The offline/test seam
    /// (`expectedCode != nil`) just clears the input.
    func resendPostcard() {
        codeInput = ""
        guard expectedCode == nil else { return }
        Task { [api, homeId] in
            _ = try? await api.request(
                HomesEndpoints.requestPostcard(homeId: homeId),
                as: RequestPostcardResponse.self
            )
        }
    }

    /// Used by the debug / preview tooling and the snapshot tests to
    /// flip between the in-transit and delivered frames without
    /// waiting on the simulated USPS clock.
    func setStage(_ next: PostcardDeliveryStage) {
        stage = next
        content = PostcardVerificationSampleData.content(for: next)
        if next != .delivered {
            codeInput = ""
        }
    }

    func verifyTapped() {
        guard primaryCTAEnabled else { return }
        Task { await verify() }
    }

    func dismissTapped() {
        pendingEvent = .dismiss
    }

    func acknowledgePendingEvent() {
        pendingEvent = nil
    }

    // MARK: - Submit

    private func verify() async {
        submitState = .submitting
        if let expectedCode {
            // Offline/test seam — compare locally, no network.
            try? await Task.sleep(nanoseconds: submitDelayNanos)
            if codeInput == expectedCode {
                submitState = .submitted
                pendingEvent = .verified(homeId: homeId)
            } else {
                submitState = .error(message: "That code didn't match. Double-check the postcard.")
                codeInput = ""
            }
            return
        }
        do {
            _ = try await api.request(
                HomesEndpoints.verifyPostcard(
                    homeId: homeId,
                    request: VerifyPostcardRequest(code: codeInput)
                ),
                as: VerifyPostcardResponse.self
            )
            submitState = .submitted
            pendingEvent = .verified(homeId: homeId)
        } catch {
            if case .transport = (error as? APIError) {
                submitState = .error(message: "You're offline. Try again when you're back online.")
            } else {
                submitState = .error(message: "That code didn't match. Double-check the postcard.")
                codeInput = ""
            }
        }
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/verify_landlord/postcard/PostcardVerificationViewModel.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.ui.screens.homes.verify_landlord.postcard

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.ui.screens.homes.verify_landlord.VerifyLandlordSubmitState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/** Nav-arg key for the home whose postcard is being verified. */
const val POSTCARD_VERIFICATION_HOME_ID_KEY: String = "homeId"

/** Lifecycle state of the physical postcard. */
enum class PostcardDeliveryStage { Mailed, InTransit, Delivered }

/** Full payload describing the postcard verification surface. */
data class PostcardVerificationContent(
    val recipientName: String,
    val street: String,
    val cityZip: String,
    val trackingNumber: String,
    val mailedOn: String,
    val inTransitOn: String?,
    val deliveredOn: String?,
    val resendAvailableOn: String,
)

object PostcardVerificationSampleData {
    val deliveredContent: PostcardVerificationContent =
        PostcardVerificationContent(
            recipientName = "Mira Patel",
            street = "412 Elm St, Apt 3B",
            cityZip = "San Francisco, CA 94114",
            trackingNumber = "#9405 5036 …8421",
            mailedOn = "Oct 9",
            inTransitOn = "Oct 11",
            deliveredOn = "Oct 12",
            resendAvailableOn = "Oct 15",
        )

    val inTransitContent: PostcardVerificationContent =
        PostcardVerificationContent(
            recipientName = "Mira Patel",
            street = "412 Elm St, Apt 3B",
            cityZip = "San Francisco, CA 94114",
            trackingNumber = "#9405 5036 …8421",
            mailedOn = "Oct 9",
            inTransitOn = "Oct 11",
            deliveredOn = null,
            resendAvailableOn = "Oct 15",
        )

    fun content(stage: PostcardDeliveryStage): PostcardVerificationContent =
        if (stage == PostcardDeliveryStage.Delivered) deliveredContent else inTransitContent

    fun stage(homeId: String): PostcardDeliveryStage =
        if (homeId.contains("delivered", ignoreCase = true)) {
            PostcardDeliveryStage.Delivered
        } else {
            PostcardDeliveryStage.InTransit
        }
}

/** Outbound events the host nav stack acts on. */
sealed interface PostcardVerificationOutboundEvent {
    data object Dismiss : PostcardVerificationOutboundEvent

    /**
     * Verify pressed and the code matched — caller should pop the
     * screen and route to the verified-home success surface.
     */
    data class Verified(val homeId: String) : PostcardVerificationOutboundEvent
}

/** Aggregate UI state for A12.7. */
data class PostcardVerificationUiState(
    val stage: PostcardDeliveryStage,
    val content: PostcardVerificationContent,
    val codeInput: String = "",
    val submitState: VerifyLandlordSubmitState = VerifyLandlordSubmitState.Idle,
) {
    val isCodeInputUnlocked: Boolean get() = stage == PostcardDeliveryStage.Delivered

    val isSubmitting: Boolean get() = submitState is VerifyLandlordSubmitState.Submitting

    val primaryCtaEnabled: Boolean
        get() = isCodeInputUnlocked && codeInput.length == CODE_LENGTH && !isSubmitting

    val primaryCtaLabel: String
        get() = if (isSubmitting) "Verifying…" else "Verify code"

    companion object {
        /** Length of the postcard code printed by the issuer. */
        const val CODE_LENGTH: Int = 6
    }
}

/**
 * View model for A12.7. Holds the current delivery stage, the user's
 * 6-char code, and a `submitState` machine identical in shape to the
 * wizard's [VerifyLandlordSubmitState].
 */
@HiltViewModel
open class PostcardVerificationViewModel
    @Inject
    constructor(
        savedStateHandle: SavedStateHandle,
    ) : ViewModel() {
        private val homeId: String =
            requireNotNull(savedStateHandle[POSTCARD_VERIFICATION_HOME_ID_KEY]) {
                "PostcardVerificationViewModel requires a '$POSTCARD_VERIFICATION_HOME_ID_KEY' nav arg."
            }

        protected open val submitDelayMillis: Long = SUBMIT_DELAY_DEFAULT_MILLIS

        /** Code printed on the postcard. Replace with the per-claim
         *  secret when the postcard issuance endpoint ships. */
        protected open val expectedCode: String = DEFAULT_EXPECTED_CODE

        private val _state =
            MutableStateFlow(
                PostcardVerificationUiState(
                    stage = PostcardVerificationSampleData.stage(homeId),
                    content =
                        PostcardVerificationSampleData.content(
                            PostcardVerificationSampleData.stage(homeId),
                        ),
                ),
            )
        val state: StateFlow<PostcardVerificationUiState> = _state.asStateFlow()
        val pendingEvent = MutableStateFlow<PostcardVerificationOutboundEvent?>(null)

        // MARK: - Mutations

        fun updateCode(raw: String) {
            val sanitized = raw.uppercase().take(PostcardVerificationUiState.CODE_LENGTH)
            _state.update { it.copy(codeInput = sanitized) }
        }

        fun resendPostcard() {
            _state.update { it.copy(codeInput = "") }
        }

        /**
         * Used by debug / preview tooling and snapshot tests to flip
         * between the in-transit and delivered frames without waiting
         * on the simulated USPS clock.
         */
        fun setStage(next: PostcardDeliveryStage) {
            _state.update { current ->
                current.copy(
                    stage = next,
                    content = PostcardVerificationSampleData.content(next),
                    codeInput = if (next != PostcardDeliveryStage.Delivered) "" else current.codeInput,
                )
            }
        }

        fun verifyTapped() {
            val snapshot = _state.value
            if (!snapshot.primaryCtaEnabled) return
            viewModelScope.launch { verify() }
        }

        fun dismissTapped() {
            pendingEvent.value = PostcardVerificationOutboundEvent.Dismiss
        }

        fun acknowledgeEvent() {
            pendingEvent.value = null
        }

        // MARK: - Submit

        private suspend fun verify() {
            _state.update { it.copy(submitState = VerifyLandlordSubmitState.Submitting) }
            if (submitDelayMillis > 0) delay(submitDelayMillis)
            val snapshot = _state.value
            if (snapshot.codeInput == expectedCode) {
                _state.update { it.copy(submitState = VerifyLandlordSubmitState.Submitted) }
                pendingEvent.value = PostcardVerificationOutboundEvent.Verified(homeId)
            } else {
                _state.update {
                    it.copy(
                        submitState =
                            VerifyLandlordSubmitState.Error(
                                "That code didn't match. Double-check the postcard.",
                            ),
                        codeInput = "",
                    )
                }
            }
        }

        companion object {
            const val SUBMIT_DELAY_DEFAULT_MILLIS: Long = 800L
            const val DEFAULT_EXPECTED_CODE: String = "4Q2K7B"
        }
    }
